File: src/TrackPartition.cpp
#include "TrackPartition.h"
#include "Track.h"
#include "WorldArea.h"
#include "Grid.h"
#include "lineMath.h"

#include <cassert>
#include <algorithm>

 TrackPartition::TrackPartition(Track &track, int relativeCellSize)
 	: track_(track), relativeCellSize_(relativeCellSize)
{
	int rows = (track_.worldArea_->bottomRight().y - track_.worldArea_->topLeft().y) / relativeCellSize + 1;
	int cols = (track_.worldArea_->bottomRight().x - track_.worldArea_->topLeft().x) / relativeCellSize + 1;
	cells.assign(rows, row_type{});
	for (auto &r : cells)
		r.assign(cols, cell_type{});
}

std::pair<int, int> TrackPartition::worldToCell(WorldPoint const& wp) const {
	int row = (wp.y / track_.grid_->cellSize() - track_.worldArea_->topLeft().y) / relativeCellSize_;
	int col = (wp.x / track_.grid_->cellSize() - track_.worldArea_->topLeft().x) / relativeCellSize_;
	if (row < 0)
		row = 0;
	if ((unsigned)row >= cells.size())
		row = cells.size()-1;
	if (col < 0)
		col = 0;
	if ((unsigned)col >= cells[0].size())
		col = cells[0].size()-1;
	return {row, col};
}

WorldPoint TrackPartition::cellToWorld(int row, int col) const {
	float x = (col * relativeCellSize_ + track_.worldArea_->topLeft().x) * track_.grid_->cellSize();
	float y = (row * relativeCellSize_ + track_.worldArea_->topLeft().y) * track_.grid_->cellSize();
	return {x, y};
}

void TrackPartition::addVertex(std::pair<int, int> index, WorldPoint const& v) {
	auto coord = worldToCell(v);
	int row = coord.first;
	int col = coord.second;
	cells[row][col].insert(index);
}

void TrackPartition::addSegment(std::pair<int, std::pair<int, int>> seg, WorldPoint const& p1, WorldPoint const& p2) {
	auto coord1 = worldToCell(p1);
	auto coord2 = worldToCell(p2);
	int row1 = std::min(coord1.first, coord2.first);
	int col1 = std::min(coord1.second, coord2.second);
	int row2 = std::max(coord1.first, coord2.first);
	int col2 = std::max(coord1.second, coord2.second);
	// we need to check every cell in the interval (r1,c1)->(r2,c2) for intersection with the line segment
	for (int i=row1; i<=row2; i++)
		for (int j=col1; j<=col2; j++) {
			auto c1 = cellToWorld(i, j);	// top-left corner
			auto c2 = cellToWorld(i, j+1);	// top-right corner
			auto c3 = cellToWorld(i+1, j);	// bottom-left corner
			auto c4 = cellToWorld(i+1, j+1);// bottom-right corner
			auto o1 = lineMath::orientation(p1, p2, c1);
			auto o2 = lineMath::orientation(p1, p2, c2);
			auto o3 = lineMath::orientation(p1, p2, c3);
			auto o4 = lineMath::orientation(p1, p2, c4);
			// now all corners must be on the same side of the line if the line doesn't intersect the cell (they must have the same non-zero orientation)
			if (o1 > 0 && o2 > 0 && o3 > 0 && o4 > 0)
				continue; // all corners are on the positive side
			if (o1 < 0 && o2 < 0 && o3 < 0 && o4 < 0)
				continue; // all corners are on the negative side
			// if we got here then the line segment intersects the current cell, we add its vertices
			cells[i][j].insert({seg.first, seg.second.first});
			cells[i][j].insert({seg.first, seg.second.second});
		}
}

std::set<std::pair<int, int>> TrackPartition::getVerticesInArea(WorldPoint topLeft, WorldPoint bottomRight) const {
	assert(bottomRight.x >= topLeft.x && bottomRight.y >= topLeft.y);
	auto coord1 = worldToCell(topLeft);
	int row1 = coord1.first;
	int col1 = coord1.second;
	auto coord2 = worldToCell(bottomRight);
	int row2 = std::min((int)cells.size()-1, coord2.first + 1);
	int col2 = std::min((int)cells[0].size()-1, coord2.second + 1);

	std::set<std::pair<int, int>> ret;
	for (int i=row1; i<=row2; i++)
		for (int j=col1; j<=col2; j++)
			std::copy(cells[i][j].begin(), cells[i][j].end(), std::inserter(ret, ret.end()));
	return ret;
}

void TrackPartition::clear() {
	for (auto &r : cells)
		for (auto &c : r)
			c.clear();
}

File: client/src/track-partition.ts
import { Grid } from "./grid";
import { lineMath } from "./math/line-math";
import { WorldPoint } from "./math/math";
import { assert } from "./utils/assert";
import { WorldArea } from "./world-area";

type element_type = [number, number];
type cell_type = Set<element_type>;
type row_type = cell_type[];
export type segment_type = [number, [number, number]];

export class TrackPartition {
	private cells: row_type[];

	/** @param relativeCellSize_: how many grid cells is one partition cell wide and high? */
	constructor(
		private readonly worldArea: WorldArea,
		private readonly grid: Grid,
		private readonly relativeCellSize_: number,
	) {
		const rows: number = (worldArea.bottomRight().y - worldArea.topLeft().y) / this.relativeCellSize_ + 1;
		const cols: number = (worldArea.bottomRight().x - worldArea.topLeft().x) / this.relativeCellSize_ + 1;
		this.cells = new Array(rows);
		for (let i = 0; i < rows; i++) {
			this.cells[i] = new Array(cols);
			for (let j = 0; j < cols; j++) {
				this.cells[i][j] = new Set<element_type>();
			}
		}
	}

	// adds a vertex into space partitioning structure
	addVertex(index: element_type, v: WorldPoint): void {
		const coord = this.worldToCell(v);
		const row = coord[0];
		const col = coord[1];
		this.cells[row][col].add(index);
	}

	// adds a line segment into the partitioning structure -> its vertices will be added to all cells that the segment overlaps
	addSegment(seg: segment_type, p1: WorldPoint, p2: WorldPoint): void {
		const coord1 = this.worldToCell(p1);
		const coord2 = this.worldToCell(p2);
		const row1 = Math.min(coord1[0], coord2[0]);
		const col1 = Math.min(coord1[1], coord2[1]);
		const row2 = Math.max(coord1[0], coord2[0]);
		const col2 = Math.max(coord1[1], coord2[1]);
		// we need to check every cell in the interval (r1,c1)->(r2,c2) for intersection with the line segment
		for (let i = row1; i <= row2; i++)
			for (let j = col1; j <= col2; j++) {
				const c1 = this.cellToWorld(i, j); // top-left corner
				const c2 = this.cellToWorld(i, j + 1); // top-right corner
				const c3 = this.cellToWorld(i + 1, j); // bottom-left corner
				const c4 = this.cellToWorld(i + 1, j + 1); // bottom-right corner
				const o1 = lineMath.orientation(p1, p2, c1);
				const o2 = lineMath.orientation(p1, p2, c2);
				const o3 = lineMath.orientation(p1, p2, c3);
				const o4 = lineMath.orientation(p1, p2, c4);
				// now all corners must be on the same side of the line if the line doesn't intersect the cell (they must have the same non-zero orientation)
				if (o1 > 0 && o2 > 0 && o3 > 0 && o4 > 0) continue; // all corners are on the positive side
				if (o1 < 0 && o2 < 0 && o3 < 0 && o4 < 0) continue; // all corners are on the negative side
				// if we got here then the line segment intersects the current cell, we add its vertices
				this.cells[i][j].add([seg[0], seg[1][0]]);
				this.cells[i][j].add([seg[0], seg[1][1]]);
			}
	}

	getVerticesInArea(topLeft: WorldPoint, bottomRight: WorldPoint): Set<element_type> {
		assert(bottomRight.x >= topLeft.x && bottomRight.y >= topLeft.y);
		const coord1 = this.worldToCell(topLeft);
		const row1 = coord1[0];
		const col1 = coord1[1];
		const coord2 = this.worldToCell(bottomRight);
		const row2 = Math.min(this.cells.length - 1, coord2[0]);
		const col2 = Math.min(this.cells[0].length - 1, coord2[1]);

		const ret = new Set<element_type>();
		for (let i = row1; i <= row2; i++) {
			for (let j = col1; j <= col2; j++) {
				for (const el of this.cells[i][j]) {
					ret.add(el);
				}
			}
		}
		return ret;
	}

	clear(): void {
		for (const r of this.cells) for (const c of r) c.clear();
	}

	// ---------------------------------- PRIVATE AREA ----------------------------------

	private worldToCell(wp: WorldPoint): number[] {
		let row: number = (wp.y / this.grid.cellSize() - this.worldArea.topLeft().y) / this.relativeCellSize_;
		let col: number = (wp.x / this.grid.cellSize() - this.worldArea.topLeft().x) / this.relativeCellSize_;
		if (row < 0) row = 0;
		if (Math.floor(row) >= this.cells.length) row = this.cells.length - 1;
		if (col < 0) col = 0;
		if (Math.floor(col) >= this.cells[0].length) col = this.cells[0].length - 1;
		return [row, col];
	}

	private cellToWorld(row: number, col: number): WorldPoint {
		const x = (col * this.relativeCellSize_ + this.worldArea.topLeft().x) * this.grid.cellSize();
		const y = (row * this.relativeCellSize_ + this.worldArea.topLeft().y) * this.grid.cellSize();
		return new WorldPoint(x, y);
	}
}
